File: 525-contiguous-array/contiguous-array.py
'''
zeros = 8
ones = 7
1   0   0   1   0   1   1   1   0   0   0   0   0   1   1

0   1   2   2   3   3   3   3   4   5   6   7   8   8   8
1   1   1   2   2   3   4   5   5   5   5   5   5   6   7

0, 0
-1, 0
0, 1
1, 2

0   1

1   1
0   1
'''

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        diffMap = {0: -1}
        res = 0

        ones, zeros = 0, 0
        for i, num in enumerate(nums):
            if num == 0:
                zeros += 1
            else:
                ones += 1
            curDiff = zeros - ones
            if curDiff in diffMap:
                res = max(res, i - diffMap[curDiff])
            else:
                diffMap[curDiff] = i
        
        return res
            

File: 525-contiguous-array/contiguous-array.java
/*
    diff = ones - zeros, i
    0   1   1   1   1   1   0   0   0
    -1  0   1   2   3   4   3   2   1
 */

class Solution {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> diffs = new HashMap<>();
        diffs.put(0, -1);
        int zeros = 0;
        int largest = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) zeros++;
            int diff = i + 1 - 2 * zeros;
            // System.out.println(diff);
            if (diffs.containsKey(diff)) {
                largest = Math.max(largest, i - diffs.get(diff));
            } else {
                diffs.put(diff, i);
            }
        }
        return largest;
    }
}